--- uart_dev_ns8250.c.orig	2013-01-30 18:49:49.000000000 +0800
+++ uart_dev_ns8250.c	2013-02-21 19:19:46.000000000 +0800
@@ -24,6 +24,8 @@
  * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
 
+#include "opt_platform.h"
+
 #include <sys/cdefs.h>
 __FBSDID("$FreeBSD: head/sys/dev/uart/uart_dev_ns8250.c 246016 2013-01-27 23:33:42Z cperciva $");
 
@@ -35,6 +37,12 @@
 #include <sys/sysctl.h>
 #include <machine/bus.h>
 
+#ifdef FDT
+#include <dev/fdt/fdt_common.h>
+#include <dev/ofw/ofw_bus.h>
+#include <dev/ofw/ofw_bus_subr.h>
+#endif
+
 #include <dev/uart/uart.h>
 #include <dev/uart/uart_cpu.h>
 #include <dev/uart/uart_bus.h>
@@ -45,6 +53,13 @@
 
 #define	DEFAULT_RCLK	1843200
 
+static int busy_detect = 0;
+
+static int broken_txfifo = 0;
+SYSCTL_INT(_hw, OID_AUTO, broken_txfifo, CTLFLAG_RW | CTLFLAG_TUN,
+	&broken_txfifo, 0, "UART FIFO has QEMU emulation bug");
+TUNABLE_INT("hw.broken_txfifo", &broken_txfifo);
+
 /*
  * Clear pending interrupts. THRE is cleared by reading IIR. Data
  * that may have been received gets lost here.
@@ -402,6 +417,21 @@
 	struct uart_bas *bas;
 	unsigned int ivar;
 
+#ifdef FDT
+	phandle_t node;
+	pcell_t cell;
+
+	/* 
+	 * Check whether uart requires to read USR reg when IIR_BUSY and 
+	 * has broken txfifo. 
+	 */
+	node = ofw_bus_get_node(sc->sc_dev);
+	if ((OF_getprop(node, "busy_detect", &cell, sizeof(cell))) > 0) {
+		busy_detect = 1;
+		broken_txfifo = 1;
+	}
+#endif
+
 	bas = &sc->sc_bas;
 
 	ns8250->mcr = uart_getreg(bas, REG_MCR);
@@ -592,6 +622,12 @@
 	bas = &sc->sc_bas;
 	uart_lock(sc->sc_hwmtx);
 	iir = uart_getreg(bas, REG_IIR);
+
+	if (busy_detect && (iir & IIR_BUSY) == IIR_BUSY) {
+		(void)uart_getreg(bas, A10_REG_USR);
+		uart_unlock(sc->sc_hwmtx);
+		return (0);
+	}
 	if (iir & IIR_NOPEND) {
 		uart_unlock(sc->sc_hwmtx);
 		return (0);
@@ -847,11 +883,6 @@
 	return (0);
 }
 
-static int broken_txfifo = 0;
-SYSCTL_INT(_hw, OID_AUTO, broken_txfifo, CTLFLAG_RW | CTLFLAG_TUN,
-	&broken_txfifo, 0, "UART FIFO has QEMU emulation bug");
-TUNABLE_INT("hw.broken_txfifo", &broken_txfifo);
-
 static int
 ns8250_bus_transmit(struct uart_softc *sc)
 {
